src/core/seat/keyboard.cpp

```diff
@@ -106,29 +108,54 @@ void wf::keyboard_t::reload_input_options()
 
     auto ctx = xkb_context_new(XKB_CONTEXT_NO_FLAGS);
 
-    /* Copy memory to stack, so that .c_str() is valid */
-    std::string rules   = this->rules;
-    std::string model   = this->model;
-    std::string layout  = this->layout;
-    std::string variant = this->variant;
-    std::string options = this->options;
-
-    xkb_rule_names names;
-    names.rules   = rules.c_str();
-    names.model   = model.c_str();
-    names.layout  = layout.c_str();
-    names.variant = variant.c_str();
-    names.options = options.c_str();
-    auto keymap = xkb_map_new_from_names(ctx, &names,
-        XKB_KEYMAP_COMPILE_NO_FLAGS);
+    struct xkb_keymap *keymap = nullptr;
 
-    if (!keymap)
+    if (!this->keymap_file.value().empty())
+    {
+        std::string path = this->keymap_file;
+        FILE *file = std::fopen(path.c_str(), "r");
+        if (file)
+        {
```

Comment:
I think it would make sense to split the two branches into separate helper functions.